boot/freeldr/freeldr/ntldr/wlmemory.c

```diff
@@ -238,6 +239,51 @@ WinLdrSetupMemoryLayout(IN OUT PLOADER_PARAMETER_BLOCK LoaderBlock)
         return FALSE;
     }
 
+    // Always map the ACPI table if present, otherwise Windows will crash.
+    if (Extension->AcpiTable)
+    {
+        PVOID AcpiTableClone = MmAllocateMemoryWithType(Extension->AcpiTableSize, LoaderFirmwarePermanent);
+        if (!AcpiTableClone)
+        {
+            ERR("Cannot allocate ACPI table\n");
+            return FALSE;
+        }
+
+        RtlCopyMemory(AcpiTableClone, Extension->AcpiTable, Extension->AcpiTableSize);
+        Extension->AcpiTable = AcpiTableClone;
+
+        union
+        {
+            PDESCRIPTION_HEADER Header;
+            PRSDT_DESCRIPTOR Rsdt;
+            PXSDT_DESCRIPTOR Xsdt;
+        } AcpiTable = {.Header = Extension->AcpiTable};
+
+        BOOLEAN is64BitsTable = _strnicmp(AcpiTable.Header->Signature, "XSDT", 4) == 0;
+        ULONG Entries = (AcpiTable.Header->Length - sizeof(DESCRIPTION_HEADER)) / (is64BitsTable ? 8 : 4);
+
+        for (ULONG i = 0; i < Entries; i++)
+        {
+            PDESCRIPTION_HEADER Entry = is64BitsTable ? 
+                    (PVOID)((ULONG_PTR)AcpiTable.Xsdt->PointerToOtherSDT[i]) :
+                    (PVOID)((ULONG_PTR)AcpiTable.Rsdt->PointerToOtherSDT[i]);
+
+            PDESCRIPTION_HEADER EntryClone = MmAllocateMemoryWithType(Entry->Length, LoaderFirmwarePermanent);
```

Comment:
This will probably waste a bunch of memory, because it allocates 1 page for each table, no matter how small it is. Why not using the heap?

Reply:
because it's the only function I know and I changed my code now it should be more efficient in terms of memory usage.